# 5. Implementing and Comparing CNN Architectures

### Task:

Convolutional Neural Networks (CNNs) can be organized into different architectures to learn features from image data.

This task implements a simplified **AlexNet architecture** and a **ResNet-like architecture** using TensorFlow/Keras to explore how different CNN architectures are structured.


In [2]:
# TASK 1: Implement AlexNet Architecture 

# Import TensorFlow/Keras for building the CNN architecture
import tensorflow as tf


# Build the AlexNet model using the layers specified in the question
alexnet_model = tf.keras.Sequential([

    # Define the expected input image shape
    # AlexNet commonly works with large RGB images, so each input contains height, width, and 3 color channels
    tf.keras.layers.Input(shape=(227, 227, 3)),

    # First convolution layer - Conv2D Layer: 96 filters, kernel size = (11,11), stride = 4, activation = ReLU
    # Uses 96 filters to begin extracting visual features from the image
    # A stride of 4 moves the filters four pixels at a time
    tf.keras.layers.Conv2D(
        filters=96,
        kernel_size=(11, 11),
        strides=4,
        activation="relu"
    ),

    # MaxPooling Layer: pool size = (3,3), stride = 2
    # Reduce the spatial size by keeping the maximum value from each 3x3 region
    tf.keras.layers.MaxPooling2D(
        pool_size=(3, 3),
        strides=2
    ),

    # Conv2D Layer: 256 filters, kernel size = (5,5), activation = ReLU
    # Second convolution layer uses 256 filters to learn more detailed features from the previous feature maps
    tf.keras.layers.Conv2D(
        filters=256,
        kernel_size=(5, 5),
        activation="relu"
    ),

    # MaxPooling Layer: pool size = (3,3), stride = 2 
    # Reduce the spatial dimensions again using Max Pooling
    tf.keras.layers.MaxPooling2D(
        pool_size=(3, 3),
        strides=2
    ),

    # Three additional convolution layers learn increasingly complex features using 3x3 filters
    # Conv2D Layer: 384 filters, kernel size = (3,3), activation = ReLU
    tf.keras.layers.Conv2D(
        filters=384,
        kernel_size=(3, 3),
        activation="relu"
    ),

    tf.keras.layers.Conv2D(
        filters=384,
        kernel_size=(3, 3),
        activation="relu"
    ),

    tf.keras.layers.Conv2D(
        filters=256,
        kernel_size=(3, 3),
        activation="relu"
    ),

    # MaxPooling Layer: pool size = (3,3), stride = 2
    # Final Max Pooling layer reduces the feature-map size before the features are passed to the Dense layers
    tf.keras.layers.MaxPooling2D(
        pool_size=(3, 3),
        strides=2
    ),

    # Flatten Layer
    # Convert the multidimensional feature maps into a one-dimensional vector for classification
    tf.keras.layers.Flatten(),

    # Fully Connected (Dense) Layer: 4096 neurons, activation = ReLU
    # First fully connected layer with 4096 neurons.
    tf.keras.layers.Dense(
        4096,
        activation="relu"
    ),

    # Dropout Layer: 50%
    # Randomly deactivate 50% of neurons during training to help reduce overfitting
    tf.keras.layers.Dropout(0.5),

    # Fully Connected (Dense) Layer: 4096 neurons, activation = ReLU
    # Second fully connected layer with 4096 neurons.
    tf.keras.layers.Dense(
        4096,
        activation="relu"
    ),

    # Apply another 50% dropout
    tf.keras.layers.Dropout(0.5),

    # Output Layer: 10 neurons, activation = Softmax
    # Output 10 class probabilities
    # Softmax converts the outputs into probabilities across the 10 possible classes
    tf.keras.layers.Dense(
        10,
        activation="softmax"
    )
])


# Print the model summary to verify the layers, output shapes, and number of model parameters
print("The Model Summary for AlexNet")
alexnet_model.summary()

The Model Summary for AlexNet


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_5 (Conv2D)               │ (None, 55, 55, 96)     │        34,944 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_3 (MaxPooling2D)  │ (None, 27, 27, 96)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_6 (Conv2D)               │ (None, 23, 23, 256)    │       614,656 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_4 (MaxPooling2D)  │ (None, 11, 11, 256)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_7 (Conv2D)               │ (None, 9, 9, 384)      │       885,120 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_8 (Conv2D)               │ (None, 7, 7, 384)      │     1,327,488 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_9 (Conv2D)               │ (None, 5, 5, 256)      │       884,992 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_5 (MaxPooling2D)  │ (None, 2, 2, 256)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 1024)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 4096)           │     4,198,400 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 4096)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 4096)           │    16,781,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 4096)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 10)             │        40,970 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 24,767,882 (94.48 MB)

 Trainable params: 24,767,882 (94.48 MB)

 Non-trainable params: 0 (0.00 B)

In [3]:
# TASK 2: Implement a Residual Block and ResNet 


# 1. Define the residual block


# i. Takes an input tensor 

# Define a function that takes:
# input_tensor = the input entering the residual block
# filters = the number of filters to use in the convolution layers
def residual_block(input_tensor, filters):

    # ii.  Applies two Conv2D layers (each with 64 filters, kernel size = (3,3), activation = ReLU)
    
    # Apply the FIRST Conv2D layer to the input tensor
    # It uses 64 filters, a 3x3 kernel, and ReLU activation
    # padding="same" keeps the height and width unchanged
    x = tf.keras.layers.Conv2D(
        filters=filters,
        kernel_size=(3, 3),
        padding="same",
        activation="relu"
    )(input_tensor)

    # Apply the SECOND Conv2D layer to the output of the first layer
    # It also uses 64 filters, a 3x3 kernel, and ReLU activation
    x = tf.keras.layers.Conv2D(
        filters=filters,
        kernel_size=(3, 3),
        padding="same",
        activation="relu"
    )(x)


    # iii. Includes a skip connection that adds the input tensor to the output before activation
    
    # Create the SKIP CONNECTION
    # Add the original input_tensor directly to the output produced by the two Conv2D layers
    x = tf.keras.layers.Add()([input_tensor, x])

    # Apply activation after the skip connection has been added, as required by the question.
    output = tf.keras.layers.ReLU()(x)

    # Return the final output of the residual block
    return output




# 2. Build a simple ResNet-like model

# Define the input for the ResNet-like model
# The model will receive RGB images with 224x224 pixels and 3 color channels
input_layer = tf.keras.layers.Input(shape=(224, 224, 3))


# i. Initial Conv2D layer

# Apply the initial Conv2D layer required by the question:
# 64 filters, a 7x7 kernel, and a stride of 2
x = tf.keras.layers.Conv2D(
    filters=64,
    kernel_size=(7, 7),
    strides=2,
    padding="same",
    activation="relu"
)(input_layer)


# ii. Apply two residual blocks

# Pass the output of the initial Conv2D layer through the first residual block defined in Step 1
x = residual_block(x, filters=64)

# Pass the output of the first residual block through
# a second residual block.
x = residual_block(x, filters=64)


# iii. Ends with a Flatten layer, Dense layer (128 neurons), and Output layer (Softmax)

# Flatten converts the feature maps produced by the residual blocks
# into a one-dimensional vector for the Dense layer
x = tf.keras.layers.Flatten()(x)

# Apply the required Dense layer with 128 neurons.
x = tf.keras.layers.Dense(
    128,
    activation="relu"
)(x)

# Create the final output layer
# Softmax converts the outputs into probabilities across 10 classes
output_layer = tf.keras.layers.Dense(
    10,
    activation="softmax"
)(x)


# Connect the input and output to create the complete ResNet-like model
resnet_model = tf.keras.Model(
    inputs=input_layer,
    outputs=output_layer,
    name="ResNet_Like_Model"
)


# Display the ResNet-like model summary as required in the expected output
print("The Model Summary for the ResNet-like Model")
resnet_model.summary()

The Model Summary for the ResNet-like Model


Model: "ResNet_Like_Model"

┏━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)        ┃ Output Shape      ┃    Param # ┃ Connected to      ┃
┡━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━┩
│ input_layer_2       │ (None, 224, 224,  │          0 │ -                 │
│ (InputLayer)        │ 3)                │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_10 (Conv2D)  │ (None, 112, 112,  │      9,472 │ input_layer_2[0]… │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_11 (Conv2D)  │ (None, 112, 112,  │     36,928 │ conv2d_10[0][0]   │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_12 (Conv2D)  │ (None, 112, 112,  │     36,928 │ conv2d_11[0][0]   │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add (Add)           │ (None, 112, 112,  │          0 │ conv2d_10[0][0],  │
│                     │ 64)               │            │ conv2d_12[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ re_lu (ReLU)        │ (None, 112, 112,  │          0 │ add[0][0]         │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_13 (Conv2D)  │ (None, 112, 112,  │     36,928 │ re_lu[0][0]       │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ conv2d_14 (Conv2D)  │ (None, 112, 112,  │     36,928 │ conv2d_13[0][0]   │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ add_1 (Add)         │ (None, 112, 112,  │          0 │ re_lu[0][0],      │
│                     │ 64)               │            │ conv2d_14[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ re_lu_1 (ReLU)      │ (None, 112, 112,  │          0 │ add_1[0][0]       │
│                     │ 64)               │            │                   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ flatten_2 (Flatten) │ (None, 802816)    │          0 │ re_lu_1[0][0]     │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_6 (Dense)     │ (None, 128)       │ 102,760,5… │ flatten_2[0][0]   │
├─────────────────────┼───────────────────┼────────────┼───────────────────┤
│ dense_7 (Dense)     │ (None, 10)        │      1,290 │ dense_6[0][0]     │
└─────────────────────┴───────────────────┴────────────┴───────────────────┘

 Total params: 102,919,050 (392.61 MB)

 Trainable params: 102,919,050 (392.61 MB)

 Non-trainable params: 0 (0.00 B)

### Observation

In Task 1, a simplified **AlexNet model** was created using several Conv2D and Max Pooling layers followed by Flatten, Dense, Dropout, and Softmax output layers. The model summary showed how the image dimensions gradually became smaller while the convolutional layers extracted features. The final model contained **24,767,882 trainable parameters**.

In Task 2, a **ResNet-like model** was created using an initial Conv2D layer followed by two residual blocks. Each residual block contained two Conv2D layers and a **skip connection**, which added the original input of the block back to the output of the convolutional layers before the final activation. The model then ended with Flatten, Dense (128), and Softmax output layers. The model summary clearly showed these skip connections through the **Add layers** and contained **102,919,050 trainable parameters**.

The main difference observed is that **AlexNet follows a mostly sequential flow from one layer to the next, while ResNet introduces skip connections that allow information to bypass convolutional layers and be added back later**. This demonstrated how different CNN architectures can organize and process image features in different ways.
